Course improvements

johnjeng opened the pull request:
Addresses #1902

Refactors the store, actions, and syncdb of the course editor out into their own files.
Fixes `Uncontrolled to controlled` warning for the checkbox on the peer grading dialogue, student email edit, and normal grading.
Fixes a bug where setting title, description, or invite email would clear the other two if none of the default values had been changed yet.
Fixes the display name of students who have yet to join.

Fixes visual of the peer grading checkbox:
**Before**:
![enable peer grading bug](https://cloud.githubusercontent.com/assets/618575/25560438/6a253a12-2d09-11e7-841f-8cec8cd957c7.png)

**After**:
![enable peer grading fix](https://cloud.githubusercontent.com/assets/618575/25560437/6a24a750-2d09-11e7-8557-1532af54df37.png)

Also adds sorting to assignments.
![sorting assignments](https://cloud.githubusercontent.com/assets/618575/25560966/2838dd04-2d16-11e7-85eb-4c831e93db58.png)


# Tested
Adding student(s) by email and name. 
Adding assignment(s) and handout(s).
Assigning/distributing assignments and handouts
Sync of course and student information across collaborators
Configuration values like the course titles and descriptions work as expected
Editing students' first name, last name, and email addresses
Sorting by student columns
Collecting, grading, peer grading, and returning assignments
Editing student project's upgrades
Creating a shared project

williamstein commented:
> Refactors the store, actions, and syncdb of the course editor out into their own files.

Wow!  I remember looking at moving actions/store out of that next function (with its crazy global state), and thinking it would be really hard.

williamstein commented:
Is this "in progress"?  It has that weird github tag, but seems ready for review.

williamstein commented:
My very first test broke things, probably related to the folder rename. 

1. Add a few students
2. Add an assignment
3. Enable peer grading with 1.
4. Assign assignment, collect, push out, then try to "Peer collect". 

**[BOOM]**  

The error includes "Somebody may have moved the folder that should have contained the assignment."

williamstein commented:
Probably not caused by the pr, but I see "warning.js?bc93:36 Warning: FormControl is changing an uncontrolled input of type text to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component." when entering grades (for the first time).

williamstein commented:
My second test breaks:
My very first test broke things, probably related to the folder rename. 

1. Add a few students
2. Add an assignment
3. Assign assignment, collect, push out, enter grades, then try to return.
**BOOM ** when returning first with:
"Try to return again:
Traceback (most recent call last):
File "/projects/10f0e544-313c-4efe-8718-2142ac97ad11/smc/src/smc_pyutil/smc_pyutil/smc_compute.py", line 1049, in <module>
main()
File "/projects/10f0e544-313c-4efe-8718-2142ac97ad11/smc/src/smc_pyutil/smc_pyutil/smc_compute.py", line 1045, in main
args.func(args)
File "/projects/10f0e544-313c-4efe-8718-2142ac97ad11/smc/src/smc_pyutil/smc_pyutil/smc_compute.py", line 915, in g
result = getattr(Project(project_id=project_id, dev=args.dev, projects=args.projects, single=args.single), function)(**kwds)
File "/projects/10f0e544-313c-4efe-8718-2142ac97ad11/smc/src/smc_pyutil/smc_pyutil/smc_compute.py", line 836, in copy_path
target_path, target_project_path))
RuntimeError: target path (=/graded-assign2) must be contained in target project path (=/projects/10f0e544-313c-4efe-8718-2142ac97ad11/smc/src/data/projects/b351affb-8b7e-4ed8-8ade-7ff3e65e9b03)"

williamstein commented:
Aside from the workflow being broken at the end, everything else I tested (most of your list) works. 

time: 26 minutes

johnjeng commented:
>The error includes "Somebody may have moved the folder that should have contained the assignment."

This happens when you assign an empty folder due to how rsync works. It happens on live but no one has complained about it probably because you never assign an empty folder.

I'm unable to replicate your second bug.

williamstein commented:
I tried again with a nonempty folder (**This is not using peer grading.**) and got this error:
```
return to student: Traceback (most recent call last):
File "/projects/10f0e544-313c-4efe-8718-2142ac97ad11/smc/src/smc_pyutil/smc_pyutil/smc_compute.py", line 1049, in <module>
main()
File "/projects/10f0e544-313c-4efe-8718-2142ac97ad11/smc/src/smc_pyutil/smc_pyutil/smc_compute.py", line 1045, in main
args.func(args)
File "/projects/10f0e544-313c-4efe-8718-2142ac97ad11/smc/src/smc_pyutil/smc_pyutil/smc_compute.py", line 915, in g
result = getattr(Project(project_id=project_id, dev=args.dev, projects=args.projects, single=args.single), function)(**kwds)
File "/projects/10f0e544-313c-4efe-8718-2142ac97ad11/smc/src/smc_pyutil/smc_pyutil/smc_compute.py", line 836, in copy_path
target_path, target_project_path))
RuntimeError: target path (=/graded-assign4) must be contained in target project path (=/projects/10f0e544-313c-4efe-8718-2142ac97ad11/smc/src/data/projects/b351affb-8b7e-4ed8-8ade-7ff3e65e9b03)
```



<img width="952" alt="screen shot 2017-05-01 at 6 19 28 pm" src="https://cloud.githubusercontent.com/assets/1276278/25600669/c3ff4298-2e9a-11e7-87bc-1b0cd1ee4ccd.png">

DrXyzzy commented:
Testing in smc-in-smc, I create a .course file and add students who already have accounts, and two projects are created for a course for each student. Below, `Hal Test2` is a student who was added to Course1 (created using current smc master) and Course2 (created from PR branch).

![2-2-dupes](https://cloud.githubusercontent.com/assets/528072/25601411/7fd46ec4-2eb1-11e7-96f3-6858162e84e6.png)

DrXyzzy commented:
Questions about placement of options in course UI:
1. Adding students happens before finding existing students, but the search box to add students is to the right of the search box to find existing students. Shouldn't the box that is used first come before (i.e. to the left of) the box to find existing?
1. Similarly, Handouts come before Assignments in logical progression of complexity. So in that sense, shouldn't the Handouts button appear to the left of Assignments in the button bar?

DrXyzzy commented:
I created a new .course file and added an existing student by searching by name, selecting the name, and clicking add. The student appears in the student list for the course, but the email does not appear. Opening and closing the course file does not cause the email to appear in the student list.

![no-email](https://cloud.githubusercontent.com/assets/528072/25602056/4925c198-2eb6-11e7-9f78-4763cbd351de.png)

DrXyzzy commented:
About renaming a course:
1. If the .course file is renamed, course title is updated to match the new file name in course settings, but student projects for the course are not renamed.
1. If a course is created, course settings is opened, and a new description is entered without changing the default title, the title box in settings is cleared:
![couse-after-rename](https://cloud.githubusercontent.com/assets/528072/25623515/d05c5742-2f1c-11e7-9625-cd88cc3aec9a.png)
1. After course description is set (see image above), shouldn't the `Edit` button be below the description rather than above?
1. What is the intent of the large `H` at upper left of the `Title and description` section of course settings (see image above)? Is that supposed to be some sort of icon?

haraldschilly commented:
ad 4: all those description headers have an icon. in that case here, it's this one: http://fontawesome.io/icon/header/ probably thinking of "title" is the "header". Good icons are important to find ones way around. Do you see a better one here: http://fontawesome.io/icons/ ?

DrXyzzy commented:
ad 4: Maybe just re-use the mortar-board icon?

DrXyzzy commented:
Ad 2: above, when the course title is cleared, the project name for each student is changed so that the title part of the project is 'undefined', e.g. `Hal Test32 - undefined`.

DrXyzzy commented:
Conclusion of review:
1. If a .ipynb file is assigned and collected, then chat is added to the right in the collected .ipynb file, then returned to the student, the chat is not visible in the graded-assignment file returned to the student project.
1. `Save` button at upper right of course file seems unnecessary. All settings changes either update immediately or have their own `Save` button.
1. Changing .course file name changes course title, but changing course title does not change .course file name.

williamstein commented:
Save button is important- don't change.  It tells you saved or if can't
save for some reason, etc


On Tue, May 2, 2017 at 11:44 AM Hal Snyder <notifications@github.com> wrote:

> *@DrXyzzy* requested changes on this pull request.
>
> Issues identified are in comments in the PR.
>
> Here's the full testing summary:
> Tests
>
>    - Peer grading visual as shown in #1903
>    <https://github.com/sagemathinc/smc/pull/1903>. ✓
>    - Adding student(s) by email and name. ✗ duplicate projects created
>    for each student; add existing student by name => student email does not
>    appear in student list
>
>
>    - Adding assignment(s) and handout(s). ✓
>    - Assigning/distributing assignments and handouts ✓
>
>
>    - Sync of course and student information across collaborators ✓ add
>    collaborator to instructor's project and all student projects become
>    visible (including the duplicates!) and .course files show same info that
>    is visible to the instructor
>    - Configuration values like the course titles and descriptions work as
>    expected ✗ Change name of .course file - changes name in settings but does
>    not change name of student project; change name in course settings does
>    change student project name; visual after edit description; upper left
>    H icon for title/descr part of settings.
>
>
>    - Editing students' first name, last name, and email addresses ✓
>    - Sorting by student columns ✓
>
>
>    - Collecting, grading, peer grading, and returning assignments ✗ chat
>    added to right of .ipynb file after assignment and collection is not
>    returned to student.
>    - Editing student project's upgrades ✗ unable to test; instructor
>    project in smc-in-smc has no upgrades to give
>    - Creating a shared project ✓
>    - Start course in old system and pickup in new system ✓
>    - Student email does not show up in student list if student already
>    has an account and you search for student by name, then select and add the
>    student. See last comment in #1649
>    <https://github.com/sagemathinc/smc/issues/1649>.
>    - ✓ Help links at upper right of course settings are great!
>    - Save button at upper right of course settings. ✗ seems unnecessary
>    since all sections either update automatically or have their own save
>    buttons. Also, shouldn't it be consistent for all sections, either
>    requiring a Save or updating as soon as the field is changed?
>    - ✗ Changing .course file name changes course title, but changing
>    course title does not change .course file name.
>    - ✓ Re-collect and re-return warn that this is a repeat action. Nice!
>    - ✓ Export grades to .csv and .py. Files created and formats check ok.
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/sagemathinc/smc/pull/1903#pullrequestreview-35856703>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABN5drNBmGTkgMNAsHRd6s6eQNOZMu5Vks5r13mdgaJpZM4NMb7x>
> .
>
-- 
Sent from my massive iPhone 6 plus.

johnjeng commented:
>"warning.js?bc93:36 Warning: FormControl is changing an uncontrolled input of type text to be controlled

Fixed b2b3284c7269cd37404c93c7c1d22abe51a63bf6

>I tried again with a nonempty folder (This is not using peer grading.) and got this error:

Fixed e1f92a2c42b103ba09e8a480dc8bf445429a8f98

>Testing in smc-in-smc, I create a .course file and add students who already have accounts, and two projects are created for a course for each student. Below, Hal Test2 is a student who was added to Course1 (created using current smc master) and Course2 (created from PR branch).

I'm unclear what you mean by this. Are you saying that `Hal Test2` has two projects for the same course? How does this even happen?

>Questions about placement of options in course UI:...

1. I assume the most common use case after adding your students is searching for them. We could actually combine these search bars into one with the placeholder text `Find or add students...` That option loses the `by name or email address` text though.
2. You may be right because instructors might always think about Handouts (like the course syllabus, text book pdfs, etc) before Assignments. But, I don't think "logical progression of complexity" is a good ordering. For example, "Shared Project" is probably the least complex of the options but it is last because that's the last thing to consider.

>I created a new .course file and added an existing student by searching by name, selecting the name, and clicking add. The student appears in the student list for the course, but the email does not appear. Opening and closing the course file does not cause the email to appear in the student list.

This is expected behavior. It prevents random people from making a course with all the "John"s on SMC, getting their emails, and spamming them.

>About renaming a course: ...

1. This actually only happens if the course title is the same as the file name. If they're different, it doesn't change it.
This is for the "common" cases
  - Create a course, say, `A2016 math481`. Notice you named it incorrectly and change the file name to `A2016 math480`). The course title gets updated automatically.
  - Add something additional to the course title. ie. the actual title instead of just the registration code. Additionally, you know your students don't care about your personal dating method and they won't take it again, but 480 is a special topics class so they might take a different 480 which would have a different title. We set the title to "Math480 - Open source mathematical software" But your file will still be named `A2016 math480` per your own naming system.
  - Basically, it's reasonable to want the to be different. Except when they're still the same. But you need a way to make them different and changing the Title in settings is the reasonable way to do that since you think about how the file fits into your scheme before what the displayed title is.
2. Fixed 10fb0172ccce7b382fa77313cfc6b26925cc7a06 -- An empty course title is a weird case. `Undefined` seems appropriate. Suggested replacement?
3. This is just our Markdown Editor. Made a discussion ticket. #1906
4. This is the same as the Title and description panel for Project Settings so I'll consider it a separate issue. Under consideration, the `H` is weird. Made a low priority ticket #1907

>Conclusion of review: ...
1. I was unable to replicate this. It did take a little longer for the chat to load.
2. Kept per @williamstein's comment.
3. Addressed above

williamstein commented:
Everything seems to work perfectly, subject to of course the constraints and other issues created above.   

One silly thing: you did the switch so "graded-" is first, but NOT the switch so "peer-grade" is first; right now it is still last.  I made https://github.com/sagemathinc/smc/issues/1910 for this.

I'm going to merge; if subtle bugs are reported on Wed, I can fix them since you're not working WED.

williamstein commented:
time: 17 minutes
